NET-444: Config wizard; user-provided ports written as strings instead of numbers

bearni95 opened the pull request:
The wizard's lack of type structure on the temporary configs allowed for the ports on the plugins to be provided as either strings or numbers. 

This PR assumes all given inputs will be strings for the ports and modifies the tests accordingly

**Enhancement:** Located and patched a bug that allowed floats to be provided as port numbers, with the according tests

bearni95 committed:
fix: added casting for string-based port input and modified relevant tests

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -164,10 +164,9 @@ Object.keys(PLUGIN_DEFAULT_PORTS).map((pluginName) => {
             return true
         },
-        default: defaultPluginPort
+        default: defaultPluginPort.toString()
```
If we wan't the end result to be a `Number`, is this reverse conversion needed? I mean, if we receive a String, we should convert it to a `Number`. But this seems to do the reverse for the default. (Also at line `183`)

bearni95 replied:
You're totally right, didn't even consider that `input` could be read as a number! 

Made the corresponding changes, no more casting up and down `numbers` and `strings`

teogeb reviewed (changes requested):
If we'd use the `number` input type instead of `input`, we could remove most of this conversion logic?

bearni95 committed:
fix: using number as input type for ports to avoid casting up and down

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -155,7 +155,7 @@ Object.keys(PLUGIN_DEFAULT_PORTS).map((pluginName) => {
             return answers.selectPlugins.includes(pluginName)
         },
-        validate: (input: string, answers: inquirer.Answers): string | boolean => {
+        validate: (input: number, answers: inquirer.Answers): string | boolean => {
```
If the input type is number, the `parseInt` is needed. But actually the type definition is wrong, I guess? Maybe worth using inquire's `filter` hook to do the conversion from string to number?

bearni95 committed:
fix: using filter/validate for number casting, tests modified accordingly

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -155,8 +155,11 @@ Object.keys(PLUGIN_DEFAULT_PORTS).map((pluginName) => {
-        validate: (input: number, answers: inquirer.Answers): string | boolean => {
-            const portNumber = parseInt(input || answers[`${pluginName}Port`])
+        filter: (input: any, answers: inquirer.Answers) => {
+            return parseInt(input || answers[`${pluginName}Port`])
```
The `answers[`${pluginName}Port`]` is always `undefined`. 

Also if an invalid number is given, the prompt will look like this:
`Select a port for the websocket Plugin [Enter for default: 7170] (7170) NaN` 
-> it is maybe impossible to input a corrected value as backspace doesn't erase the `NaN`part`

At lines 190 and 194 we don't need `parseInt` as the value is already a `number`. Maybe it would make sense the define a custom interface for `answers` to document the answer types and get warnings if there are any type annotation mismatches?

bearni95 replied:
The `filter` method is the one interfering with the correction of values, pushing a fix without `filter` nor the crazy amount of casting we used to have

teogeb replied:
Did this solve the `Select a port for the websocket Plugin [Enter for default: 7170] (7170) NaN`, or that `answers[${pluginName}Port]` is always undefined. Also isn't the `input` type of `filter` known?

bearni95 committed:
fix: filter fn was blocking re-introduction of wrong values

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -185,11 +184,11 @@ export const getConfigFromAnswers = (answers: inquirer.Answers): any => {
                     // the publishHttp plugin is special, it needs to be added to the config after the other plugins
                     config.httpServer = {
-                        port: answers[`${pluginName}Port`]
+                        port: parseInt(answers[`${pluginName}Port`])
```
If the same modification is needed for two places, it is maybe ideal to extract the value to a separate variable so that the logic is in one place

bearni95 committed:
fix: refined types and casting for ports, selectDestinationPath

bearni95 committed:
fix: cast port using Number() rather than parseInt() to prevent float values to be interpreted as valid integers

bearni95 committed:
fix: move duplicate fn call into a const

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -208,7 +206,7 @@ export const selectDestinationPathPrompt = {
     validate: (input: string, answers: inquirer.Answers = {}): string | boolean => {
         try {
-            const filePath = input || answers.selectDestinationPath
+            const filePath = input
```
Could remove the redundant variable?

teogeb commented on packages/broker/src/ConfigWizard.ts:
```diff
@@ -155,16 +155,15 @@ Object.keys(PLUGIN_DEFAULT_PORTS).map((pluginName) => {
-        validate: (input: string, answers: inquirer.Answers): string | boolean => {
-            const portNumber = parseInt(input || answers[`${pluginName}Port`])
+        validate: (input: string | number): string | boolean => {
+            const portNumber = (typeof input === 'string') ? Number(input) : input
```
This change solved a bug, which happened if input has decimal fractions, e.g. `1234.56`. Worth mentioning in the PR description and maybe also to add a test so that we are protected from that in the future.

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -146,7 +146,7 @@ describe('ConfigWizard', () => {
 
         it ('should exercise the plugin port assignation path', () => {
-            const port = '3737'
+            const port = 3737
```
Currently `getConfigFromAnswers` expects that the port can be a `number` or a `string`. It could test the both situations.

teogeb commented:
If user gave input `7170` to the `websocket` port, it shouldn't be stored to the config file as it is the default port. Currently it stores it:
```
"websocket": {
  "port": 7170
 }
```

If user just accepts the default, it works ok and doesn't store the value.

bearni95 committed:
fix: round of comments:

- removed redundant variable filePath
- added test coverage for float point numbers
- updated tests for port coverage typings
- fixed bug where user-given default ports overwrote them

bearni95 commented:
> If user gave input `7170` to the `websocket` port, it shouldn't be stored to the config file as it is the default port. Currently it stores it:
> 
> ```
> "websocket": {
>   "port": 7170
>  }
> ```
> 
> If user just accepts the default, it works ok and doesn't store the value.

Should be accounted for in the last commit

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -31,15 +31,21 @@ describe('ConfigWizard', () => {
+        it ('invalid data: float-point number', () => {
+            const validate = portPrompt.validate!
+            const port = 55.55
+            expect(validate(port)).toBe(`Non-numeric value provided`)
```
Bit misleading as 55.55 is numeric value, it is just not an integer

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -31,15 +31,21 @@ describe('ConfigWizard', () => {
         it ('happy path', () => {
             const validate = portPrompt.validate!
-            expect(validate('7070')).toBe(true)
+            expect(validate(7070)).toBe(true)
```
The validate method still supports the string input, could keep the original test and add this string test

bearni95 committed:
fix: added validation distinguishment between NaNs and non-integers;added validation test for ports on happy path using string port as well as numeric

bearni95 committed:
fix: extracted common port validation functionality into assertValidPort method
